[ch6494] Update `pgo scaledown --query` / `pgo failover --query`

jkatz opened the pull request:
**Checklist:**

 <!--- Make sure your PR is documented and tested before submission. Put an `x` in all the boxes that apply: -->
 - [x] Have you added an explanation of what your changes do and why you'd like them to be included?
 - [x] Have you updated or added documentation for the change, as applicable?
 - [x] Have you tested your changes on all related environments with successful results, as applicable?



**Type of Changes:**

 <!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
 - [x] Bug fix (non-breaking change which fixes an issue)
 - [x] New feature (non-breaking change which adds functionality)
 - [x] Breaking change (fix or feature that would cause existing functionality to change)



**What is the current behavior? (link to any open issues here)**

The original genesis of this was to fix a bug that was shown in 053387c to ensure that a replica that was not currently matched with a primary could still have its information displayed without breaking the `pgo scaledown --query` and `pgo failover --query` commands.

However, this turned into an inspection of the information that these commands are returning and it was decided to make it a bit more helpful.

The commands provided information that could help a
user make a decision about which instance in a PostgreSQL cluster they
would want to fail over to, it could be challenging to decipher as not
everyone is versed in PostgreSQL log sequence numbers (LSN).

**What is the new behavior (if this is a feature change)?**

This updates the output for both `pgo scaledown --query` and `pgo failover --query` to give the user information that can help them make a better decision around these actions.

This changes leverages the information that is being collected by the
new HA system in order to provide better feedback about the status of
the replicas available. This information includes:

- NAME: the name of the replica
- STATUS: the current status of the replica
- NODE: the node that the replica resides on
- REPLICATION LAG: how far behind the replica is behind the primary

Additionally, "TIMELINE" is available in the API but is not currently
displayed to the user.

This also cleans out some now unused code.

Issue: [ch6494]

**Other information**:

The original bug fix is still kept as part of this commit in case the other changes are rejected.

This does make modifications to the API returns for the `--query` commands on `scaledown` + `failover`.

Issue: [ch6494]

abrightwell commented on pgo/cmd/scaledown.go:
```diff
@@ -76,31 +76,48 @@ func init() {
+// flag is specified
 func queryCluster(args []string, ns string) {
 
+	// iterate through the clusters and output informatio about each one
```
`information`

abrightwell commented on pgo/cmd/scaledown.go:
```diff
@@ -76,31 +76,48 @@ func init() {
 	for _, arg := range args {
+		// indicate which cluster this is. Put a newline before to put some
+		// separation between each line
+		fmt.Println(fmt.Sprintf("\nCluster: %s", arg))
```
`fmt.Printf(...)`?

abrightwell commented on util/failover.go:
```diff
@@ -129,169 +116,93 @@ func GetPod(clientset *kubernetes.Clientset, deploymentName, namespace string) (
+//
+// Statistics include: the current node the replica is on, if it is up, the
+// replication lag, etc.
+func ReplicationStatus(request ReplicationStatusRequest) (ReplicationStatusResponse, error) {
```
I'm not sure how I feel about wrapping the 'client' handles in the request. I'm understanding that this is simply a util function, but it seems like the request should be 'data' only, IMV. :thinking: Based on what it's doing, seems like this could be it's own or part of a related 'service' and it would then have/be aware of it's own client state?

jkatz replied:
@abrightwell I think this is something to think about for a future release. My goal was to take what we have in the existing infrastructure and ensure it works for the new HA changes. To have it work in that way would require a much larger effort compared to the remainder of the HA work that needs to take place.

That said, I could be convinced that this function could leave in the `apiserver` realm given it's called primarily by two `apiserver` endpoints, which would be a significantly lower effort to take place.

abrightwell replied:
> I think this is something to think about for a future release. My goal was to take what we have in the existing infrastructure and ensure it works for the new HA changes. To have it work in that way would require a much larger effort compared to the remainder of the HA work that needs to take place.

:+1:

abrightwell commented on util/failover.go:
```diff
@@ -129,169 +116,93 @@ func GetPod(clientset *kubernetes.Clientset, deploymentName, namespace string) (
 
-	defer conn.Close()
+	// We need to create a quick map of "instance name" => node name
+	// We will iterate through the pod list once to extract the name we refere to
```
`refer`

abrightwell commented on util/failover.go:
```diff
@@ -129,169 +116,93 @@ func GetPod(clientset *kubernetes.Clientset, deploymentName, namespace string) (
-}
 
-func getSecrets(clientset *kubernetes.Clientset, cluster *crv1.Pgcluster, namespace string) ([]msgs.ShowUserSecret, error) {
+		// get the instance name that is recognize by the Operator, which is the
```
`recognized`

abrightwell commented on util/failover.go:
```diff
@@ -129,169 +116,93 @@ func GetPod(clientset *kubernetes.Clientset, deploymentName, namespace string) (
-func getSecrets(clientset *kubernetes.Clientset, cluster *crv1.Pgcluster, namespace string) ([]msgs.ShowUserSecret, error) {
+		// get the instance name that is recognize by the Operator, which is the
+		// first part of the name
+		instance.Name = instanceNamePattern.FindStringSubmatch(rawInstance.PodName)[1]
```
FWIW, it seems like this might be worthy of being it's own utility function, something like `GetInstanceName(...)`?

jkatz replied:
This is very specific in deriving the instance name from the Pods that is being output by Patroni. If we need it some more, we could consider moving it into its own utility function.

abrightwell replied:
Sounds good.

abrightwell commented on util/failover.go:
```diff
@@ -344,6 +255,27 @@ func ToggleAutoFailover(clientset *kubernetes.Clientset, enable bool, pghaScope,
 	return nil
 }
 
+// createInstanceNodeMap creates an mapping between the names of the PostgreSQL
```
`a`

abrightwell commented on util/failover.go:
```diff
@@ -344,6 +255,27 @@ func ToggleAutoFailover(clientset *kubernetes.Clientset, enable bool, pghaScope,
+func createInstanceNodeMap(pods *v1.PodList) map[string]string {
+	instanceNodeMap := map[string]string{}
+
+	// Iterate through each pod that is return and get the mapping between the
```
`returned` ... `them`

jkatz commented on pgo/cmd/failover.go:
```diff
@@ -93,39 +93,59 @@ func createFailover(args []string, ns string) {
 
+	// indicate which cluster this is. Put a newline before to put some
+	// separation between each line
+	fmt.Println(fmt.Sprintf("\nCluster: %s", args[0]))
```
Maybe...but I need a newline anyway. I could consolidate though to make it clearer.

jkatz commented:
@abrightwell Thanks for the review. I addressed most of the comments.

The point about the `util.ReplicationStatus` status mixing the client request information is valid, I believe the question is do we want to do another lift to move it into the `apiserver` realm. It appears that this function is only used there, so it would not be complicated to move it into an area it can be commonly accessed from.

OTOH the goal was to keep it as much of a "drop in replacement" from the previous function, and it is not the only function in there that uses portions of the client request.

Open either way on this one.

abrightwell commented:
If the original intent was to keep it a 'drop in', then I think as it sits is fine. Especially since there is already precedent for it. And FWIW, the boy scout in me would want to clean up the rest along with it and that's just unnecessary scope creep. Keeping the changes relevant and minimal given current goals I think is the right choice for now.
